Issue #2266: Fixed blank line indentification for EmptyLineSeparator check

pbaranchikov opened the pull request:
Current implementation count everything outside AntLR tree as blank lines. Javadoc and comments among them.

This patch checks, that lines between tokens are actually empty (trim().isEmpty()).

Checkstyle reports are uploaded to [my site] (http://pbaranchikov.github.io/checkstyle/empty-line-separator/checkstyle.html)
For guava project, found 1591 violations against 1576 for Checkstyle 6.11.2.

Validations are also fixed in Checkstyle code itself

mkordas commented:
@pbaranchikov, thanks!

1) Can you provide the diff between report from 6.11.2 and your changes? It would be much easier to focus just on differences.
2) Can you fix the violations in Checkstyle in separate commit for this issue?

romani commented:
Separate commit for fix is required to let us easily revert your functional changes if blocking problem is found and keep your formatting changes as they are valid no matter what.

romani commented:
Reports link is redirecting to non html report, please recheck.

pbaranchikov commented:
Splitted the commit into 2 ones. Fixed link to the report
[report for checkstyle-6.11.2](http://pbaranchikov.github.io/checkstyle/empty-line-data/checkstyle-result-orig.xml)
[report for this pull request](http://pbaranchikov.github.io/checkstyle/empty-line-data/checkstyle-result-patched.xml)
[diff between reports](http://pbaranchikov.github.io/checkstyle/empty-line-data/checkstyle-result.diff)
Fixed the full-site report [link](http://pbaranchikov.github.io/checkstyle/empty-line-separator/checkstyle.html)

mkordas commented:
@pbaranchikov
Diff looks very good (although not in every case context is enough to locate file with change), code looks fine for me too. @romani, you can take over.

romani commented:
@pbaranchikov , please provide HTML reports, with test project sources copied in report.
XML files are so hard to validate, I can do diff myself in GUI merger (Meld). I expect from you 
confirmation that all changes are valid.

here is a case:
```
<file name="/home/shared/home/pavel/apps/checkstyle/contribution/checkstyle-tester/src/main/java/guava/guava-gwt/src-super/com/google/common/collect/super/com/google/common/collect/Lists.java">
<error line="534" severity="error" message="&apos;METHOD_DEF&apos; should be separated from previous statement." source="com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck"/>
```
I do not see a file: 
https://github.com/google/guava/tree/master/guava-gwt/src-super/com/google/common/collect/super/com/google/common/collect

just provide HTML reports, and do analysis of diff yourself.

pbaranchikov commented:
@romani , you do not see Lists in the reporsitory of Guava, because, checkstyle-tester uses tag 18.0 of Guava to test on. Take a look at tags in Guava repository.

New reports for [my patch] (http://pbaranchikov.github.io/checkstyle/empty-line-separator/checkstyle.html) and [checkstyle-6-11.2] (http://pbaranchikov.github.io/checkstyle/empty-line-orig/checkstyle.html)

romani commented:
diff: https://www.diffchecker.com/l0tpaz3r
only new violations, all violations are valid, good job!

One minor item to change, before merge:

> private boolean hasEmptyLineAfter(DetailAST token) {
> // If there are lines between current and next token, check if there are empty ones

comments over a block a first sign that you need to make a function from it - please do.
As you make it as function, you will make methods as with single exit point - that is also need to be done.

Right now we enforce limit of RETURN as "2" , but that is transitional limit, eventually we will have limit "1".

pbaranchikov commented:
@romani:
* reduced number of returns to 1
* extracted method to detect blank line existance in line range

pbaranchikov commented:
@romani, could you please restart build for this pull request
[This](https://travis-ci.org/checkstyle/checkstyle/jobs/85949185) seem to be an error in git, not in build.

romani commented:
Restarted

romani commented:
thanks a lot !

merged as FF.
